app/lib/clients/vault/key_value.rb

```diff
@@ -4,20 +4,35 @@ module KeyValue
       extend Policy
 
       def kv_read(identity, path)
-        verify_policy(identity, policy_path(path))
+        s = Secret.find_by(path: path)
+        verify_policy(identity, producer_policy_path(path), s&.read_groups, consumer_policy_path(path))
         client.kv(kv_mount).read(path)
       end
 
-      def kv_write(identity, path, data)
-        create_kv_policy(path)
-        assign_policy(identity, policy_path(path))
-        client.logical.write("#{kv_mount}/data/#{path}", data: data)
+      def kv_write(identity, read_groups, path, data)
+        # only producer can replace existing secret
+        if client.kv(kv_mount).read(path)
+          verify_policy(identity, producer_policy_path(path))
+        end
+
+        create_kv_policies(path)
+        assign_identity_policy(identity, producer_policy_path(path))
+        assign_groups_policy(read_groups, consumer_policy_path(path))
+        s = client.logical.write("#{kv_mount}/data/#{path}", data: data)
+        Secret.find_or_create_by(path: path).update(owner: identity.sub, read_groups: read_groups)
```

Comment:
Would it be better to write the metadata first?  (i.e. line 22 before line 21)?

Reply:
if creating secret in Vault fails do we want the metadata?

Reply:
I was thinking it is better to have the metadata without the secret, than the reverse.